pkg/storaged/btrfs/subvolume.jsx

```diff
@@ -242,38 +245,115 @@ function subvolume_delete(volume, subvol, mount_point_in_parent, card) {
+                fstab_subvol.fake_id = fake_id++;
+                subvols_by_pathname[fstab_subvol.pathname] = fstab_subvol;
+                subvols.push(fstab_subvol);
+            }
+        }
+    }
+
+    // Find parents
+    for (const pn in subvols_by_pathname) {
+        let dn = pn;
+        while (true) {
+            dn = dirname(dn);
+            if (!dn) {
+                subvols_by_pathname[pn].parent = 5;
+                break;
+            } else if (subvols_by_pathname[dn]) {
+                subvols_by_pathname[pn].parent = subvols_by_pathname[dn].fake_id;
+                break;
+            }
+        }
+    }
+
+    return subvols;
+}
+
+export function make_btrfs_subvolume_pages(parent, volume) {
+    let subvols = client.uuids_btrfs_subvols[volume.data.uuid];
+    if (!subvols) {
+        const block = client.blocks[volume.path];
+        subvols = get_subvols_from_fstab(block.Configuration);
+    }
+
+    const root = subvols.find(s => s.id == 5);
+    if (root)
+        make_btrfs_subvolume_page(parent, client.blocks[volume.path], volume, root, "", subvols);
+}
+
+export function make_btrfs_subvolume_pages_from_child_config(parent, backing_block) {
+    const block_crypto = client.blocks_crypto[backing_block.path];
+    const subvols = get_subvols_from_fstab(block_crypto.ChildConfiguration);
+
+    const root = subvols.find(s => s.id == 5);
+    if (root)
+        make_btrfs_subvolume_page(parent, backing_block, null, root, "", subvols);
+}
+
+function make_btrfs_subvolume_page(parent, backing_block, volume, subvol, path_prefix, subvols) {
     const actions = [];
 
-    const use = btrfs_usage(client, volume);
-    const block = client.blocks[volume.path];
-    const fstab_config = get_fstab_config_with_client(client, block, false, subvol);
+    const content_block = volume && client.blocks[volume.path];
+    const fstab_config = get_fstab_config(backing_block, true, subvol);
     const [, mount_point, opts] = fstab_config;
     const opt_ro = extract_option(parse_options(opts), "ro");
-    const mismount_warning = check_mismounted_fsys(block, block, fstab_config, subvol);
-    const mounted = is_mounted(client, block, subvol);
+    const mismount_warning = check_mismounted_fsys(backing_block, content_block, fstab_config, subvol);
+    const mounted = content_block && is_mounted(client, content_block, subvol);
     const mp_text = mount_point_text(mount_point, mounted);
     if (mp_text == null)
         return null;
     const forced_options = [`subvol=${subvol.pathname}`];
-    const mount_point_in_parent = get_mount_point_in_parent(volume, subvol);
+    const mount_point_in_parent = volume && get_mount_point_in_parent(volume, subvol);
 
     if (client.in_anaconda_mode()) {
         actions.push({
             title: _("Edit mount point"),
-            action: () => edit_mount_point(block, forced_options, subvol),
+            action: () => edit_mount_point(content_block || backing_block, forced_options, subvol),
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20070-20240305-074754-13ce2d6f-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)